_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,161 @@
+---
+layout: post
+published: true
+title: 'Optimizer Bug Causing the Removal of Storage Writes Before Calls to Functions Conditionally Terminating the Transaction'
+date: '2022-09-07'
+author: Solidity Team
+category: Security Alerts
+---
+On September 5, 2022, a bug in Solidity's Yul optimizer was found by differential fuzzing.
+
+The bug was introduced in version 0.8.13 and [Solidity version 0.8.17](https://github.com/ethereum/solidity/releases/tag/v0.8.17), released on September 07, 2022, provides a fix. The bug is significantly easier to trigger with optimized via-IR code generation, but can theoretically also occur in optimized legacy code generation.
+
+We assigned the bug a severity of "medium/high".
+
+
+## Which Contracts are Affected?
+
+A prerequisite for a contract to be affected is the use of ``return(...)`` or ``stop()`` in an inline assembly block (note that ``return(...)`` here refers to the builtin inline assembly function, not to the ``return`` statement of Solidity). Such an inline assembly call will not return from the current function, but instead result in an early successful (i.e. non-reverting) termination of the entire external EVM call.
+
+If your contract doesn't contain any inline assembly block with ``return(...)`` or ``stop()`` statements, you're unaffected.
+
+If it does and this early termination happens conditionally within a function, the optimizer could incorrectly remove storage writes before calls to the function.
+
+More concretely, the bug can occur if a contract contains the following pattern:
+
+1. A storage write.
+2. A call to a function that conditionally terminates using inline assembly as described above, but may also regularly return to the caller.
+3. Either of the following:
+   a) Another (unconditional) storage write overwriting the write in (1).
+   b) An unconditional revert.
+
+With legacy code generation, all of this, i.e. steps 1-3, need to happen in a single inline assembly block and the function call in (2) needs to be a call to a user-defined assembly function (see below for examples of affected Yul snippets). Also note that during legacy code generation, the Yul optimizer is only run on inline assembly snippets that do not refer to Solidity variables, further reducing the amount of potentially affected cases.
+
+However, when using via-IR code generation the entire contract is first translated to Yul and then optimized as a whole. In this case, 1-3 can happen in solidity code and only the function called in (2) needs to involve an inline assembly block calling ``return(...)`` or ``stop()``. Note that due to inlining, this inline assembly block can also merely occur in another nested function call, i.e. any call in (2) that can both return to the caller and terminate via assembly is susceptible.
```

Comment:
It does say that 1 through 3 can all happen in Solidity :-). I can remark that the inline assembly block in 2 is required, since the compiler won't generate return or stop otherwise.